README

                                                             mddb, Version 1.3.2
	           	                                 -----------------------
		                                            The CyberSKA Project
		                                                   K.R. Anderson
		                                                      2012-12-12


Introduction
------------

The mddb (metadata database) package is designed to receive, analyze,
validate, and to the extant practicable, remediate metadata as presented in a
FITs-like header file, usually one that would be produced by the CyberSKA metaData
package. This effort is undertaken in order to then present the remediated
metadata to CADC's fits2caom package, which receives these data and inserts
them into the CADC CAOM database (Common Astronomical Observation Model). The
mddb package can handle and process two (2) of the four (4) dataset formats
that are accomodated by the CyberSKA portal and the metaData package: FITS
images and CASA Images. This is due to the fact that the data model described
by CAOM v1.0 accomodates imaging data, and is not capable of modelling radio
visibility. CAOM v2.0 is expected to be able to accomodate Visibility data in
a future release. The CAOM database at CADC then provides exposure to VO
services, and CyberSKA data become available through the VAO.

mddb provides a library of classes, one each for each kind of dataset
format/type as described by the metaData package, a library of utility
functions and a command line interface for easy execution on a particular dataset.

Classes
-------
The main classes (in "mddb.mains") are presented below, and named in an
obvious way.  Each class is presented, java-like, in its own similarly named module:

	BaseDbInsert
	DbCImageInsert
	DbFitsInsert
	DbMSInsert
	DbUVFitsInsert

Implementation is done in the subclasses, which inherit all public methods from the
BaseDbInsert class. One public method, configureOverride(), is abstract and
*must* be implemented in any subclasses. The configureOverride() method then
calls the implemented data engineering, and nominally performed by private
methods implemented in the subclass.

				      BaseDbInsert
				            |
	      ----------------------------------------------------------
              |                  |                  |                  |
	DbCImageInsert	   DbFitsInsert		DbMSInsert	DbUVFitsInsert

Environment
-----------
The mddb package requires, and is dependent upon, the directory structure
known as "DQS".  This directory has the following base structure,

      $DQS/:
      ---
       CONFIG/
       DATASETS/
       fits2caom/
       HEADERS/
       logs/
       mddb.cfg
       OVERRIDE/
       VALIDATE/

Users should acquire this directory as a tar archive and deposit it on their
system.  Once in place, an environment variable, $DQS, should be defined to
indicate this location. Users will then need to edit the provided mddb.cfg
file and set the internal 'dqs' DEFAULT variable to reflect their $DQS env var.
(configs do not interpolate shell variables). Most directories will be empty
at first, though CONFIG/ and fits2caom/ must provide the expected contents.
Furthermore, a 'mddb.cfg' file is also provided in the mddb package and can be
found in mddb/config/ .

Basic operations
----------------
mddb first recieves a header file, which may be located in $DQS/HEADERS, in
which case, no path need be prepended to the filename. If in another location,
the header file first will be copied to $DQS/HEADERS to retain a copy. The
configuration process will then commence and produce a .override file, placed
in $DQS/OVERRIDE/, and named as the file name head + 'override'. This is the
set of metadata that is actually presented to the CAOM database. A validation
process then confirms the override file and whether it is suitable for
database insertion. A user may see one of two kinds of files written at this
stage:

	fileheader.nogo      or
	fileheader.valid

The validation process will write such files to the $DQS/VALIDATE directory.
A .nogo file indicates that an unsupported MIME-type was passed, which in the
case of this current version of mddb, would be either 'image/ms-uvw' or
'image/fits-uvw.' If the header file produces a .valid file, this file may 
contain up to three (3) levels of messaging:

	INFO
	WARN
	ERROR

The presence of any one 'ERROR' message will prevent the fits2caom database
insertion. These .valid files should be used as feedback to the data provider
as to what exactly is wrong with the metadata that prevents CAOM database
insertion.

An example .valid file showing the three (3) levels of messages

   27200.valid:
   ------------
   # MDDB Validation report for /srv/cyberska/DQS/OVERRIDE/27200.override
   # Written 2012-12-11
   # Package Name: mddb
   # Package Version: 1.3.2

   ERROR: [N]EXTEND keyword(s) detected. Cannot insert MEF files
   INFO:  Metadata release date set: 2099-01-01
   ERROR: Unrecognized FITS projection code:: NCP
   ERROR: Unrecognized FITS projection code:: NCP
   WARN:  INSTRUME set to ATCA
   INFO:  Wrote EQUINOX: 2000.0
   INFO:  Wrote RADESYS: FK5
   INFO:  WCS Rotation angle: 0.0
   INFO:  Wrote CD Matrix
   INFO:  Data URI http://dms.cyberska.org:8080/dlmanager/getfile?fileid=24724

Even though a single ERROR message will prevent database insertion, the
configuration and confirmation processes are run in their entirety in order to
provide as complete a summary of metadata issues as possible. It is imagined
that this .valid file may be sent to the data provider to help correct a
dataset's metadata.

Provided that no ERROR level messages appear in the file header's .valid file,
the fits2caom process is then executed.  This still may fail for various
reasons. For example, an unrecognized coordinate CTYPE such as 'RM', which
cannot be accomodated by the CAOM data model. However, the mddb package has
exercised a great deal of effort to this point in order to minimize the likelihood
of fits2caom failure. Incorrect or unaccommodated WCS metadata will
be the mostly likely source of failure, both at the .valid file level and in
fits2caom execution.

Interfaces
----------
1) Command line interface
mddb provides a command line interface, dbinsert, which will nominally accept
three (3) arguments, one of which will be the filename of the metadata header
file for the dataset to be inserted into the CAOM database.

  Usage:

  $ mddb/dbinsert

    dbinsert running mddb v1.3.2

    Usage: dbinsert [--help] [--public]
    [--verbose][--nodb][--config=<DQS-config-file>] \
    --mtype=<mime-type> --hdr=<hdrfile> --uri=<uri> 

    Three (3) keyword arguments are required. 

    --mtype= 	<mime-type> of dataset to be inserted
		One (1) of CyberSKA metadata mime-types:

			image/fits-image, 
			image/ms-image, 
			image/fits-uvw,
			image/ms-uvw

    --hdr=   	<header file for the dataset>
    --uri=   	<URI to CyberSKA platform page, or download link>,
		which should look like,

		http://www.cyberska.org/pg/file/read/<nnnnn>, or
		http://dms.cyberska.org:8080/dlmanager/getfile?fileid=<nnnnnn>

    --public	The dataset metadata are publicly available.
    --verbose	Turn on stdout messages.
    --nodb	No DB insert. executeInsert() is not called.

    A user may pass a DQS configuration file, i.e., an mddb.cfg:

    --config= 	<config-file> path to a DQS config file.

		If "--config" is not provided at the command line, a search 
		for "mddb.cfg" will occur on ~ and $DQS.  An Exception is
		raised if an mddb.cfg file cannot be found.

    --help 	This message.


2) API
Programmatically, header metadata validation and database insertion can be
performed by instatiating objects from the classes appropriate to the dataset
in question and then calling the available public methods.

Eg.,

  The rUtils.run() function provide typical usage of these classes and method
  calls (variables passed as arguments are named to manifest meaning).

  'image/fits-image':

	dbtask  = dbFitsInsert.DbFitsInsert(configFile, verbose)
        override= dbtask.buildOverride(hdrFile, uri, public)
        msgTypes= dbtask.confirmOverride(override, fileType)
	dbtask.executeInsert(collid, override, uri)


  'image/ms-image': 

  	dbtask  = dbCImageInsert.DbCImageInsert(configFile, verbose)
        override= dbtask.buildOverride(hdrFile, uri, public)
        msgTypes= dbtask.confirmOverride(override, fileType)
        dbtask.executeInsert(collid, override, uri)

  (The following are here for example purposes, but readers now understand
   that these types will cause NotImplementedError exceptions to be raised.)

  'image/fits-uvw':

  	dbtask  = dbUVFitsInsert.DbUVFitsInsert(configFile, verbose)
        override= dbtask.buildOverride(hdrFile,uri,public)   -- NotImplementedError!
        msgTypes= dbtask.confirmOverride(override,fileType)
        dbtask.executeInsert(collid,override,uri)

  'image/ms-uvw': 

        dbtask  = dbMSInsert.DbMSInsert(configFile, verbose)
        override= dbtask.buildOverride(hdrFile,uri,public)  -- NotImplementedError!
        msgTypes= dbtask.confirmOverride(override,fileType)
        dbtask.executeInsert(collid,override,uri)



-----------------



A:: The mddb package

There are two components to the mddb package. mddb python package itself,
and the DQS (Data Query Service) directory structure.

1) mddb is a pure python package. It has no dependencies on third party packages.
The contents will appear in this form:

mddb/
  ChangeLog
  README (This file)
  __init__.py
  config/
  convert/
  db/
  dbinsert --> dbinsert.py
  dbinsert.py
  docs/
  mains/
  mddbVersion.py
  scripts/
  utils/


subdirectories, also importable packages:

config/
  __init__.py
  mddb.cfg
  mddbEnv.py

convert/
  converters.py
  __init__.py
  converters.py
  keymaps.py
  mjdConversions.py

docs/
  baseDbInsert.html
  config.html
  config.mddbEnv.html
  convert.converters.html
  convert.html
  convert.keymaps.html
  convert.mjdConversions.html
  dbCImageInsert.html
  db.confirmKeys.html
  dbFitsInsert.html
  db.html
  dbinsert.html
  dbMSInsert.html
  dbUVFitsInsert.html
  db.xDBKeys.html
  mddbVersion.html
  utils.f2cArgs.html
  utils.html
  utils.rUtils.html

db/
 confirmKeys.py
 __init__.py
 checkSetKeys.py
 confirmKeys.py
 xDBKeys.py

mains/
  __init__.py
  baseDbInsert.py
  dbCImageInsert.py
  dbFitsInsert.py
  dbMSInsert.py
  dbUVFitsInsert.py

scripts/
 checkSetKeys.py
 keyAssertions.py
 public_uris.py
 public_uris_fullset.py
 testrun.py

utils/
 __init__.py
 f2cArgs.py
 keyAssertions.py
 rUtils.py

2) The DQS directory structure, which mddb uses for various input and output files.
It will arrive as a tarball and ravel into this form

 /srv/cyberska/DQS:
   CONFIG/
   DATASETS/
   fits2caom/
   HEADERS/
   logs/
   mddb.cfg
   OVERRIDE/
   runf2c.sh
   scripts/
   uris/
   VALIDATE/

Users will generally not need to consult or interact with the DQS directory at all.
The mddb package will be the primary mangler of the DQS directory.


B:: Installation
_________

1) Untar the mddb tarball and install the mddb package wherever you like.
   This path must be added to the user's PYTHONPATH in order that imports are properly done.
   The command line interface is provide through the executable module, dbinsert.py,
   which is sym-linked by the straight-up dbinsert. Users should add the location of dbinsert
   to their $PATH environment variable.

2) Untar the DQS tarball in a preferred location.  Users must set an environment variable $DQS
   to be this location. 

eg.,

   export DQS=/srv/cyberska/DQS

or if you are silly enough to use csh

     setenv DQS /srv/cyberska/DQS

$DQS is critical to the operational environment of mddb and must be set.  Essentially, it is the
     operational environment of mddb.



C:: Getting things running.
_________

1) The .dbrc file.
    Users must have a .dbrc file in their home directories.
    This is the rc file that fits2caom uses to establish the database connection.
    It will have the form

    DEVSYBASE cyberska <username> <passwd> com.sybase.jdbc3.jdbc.SybDriver jdbc:sybase:Tds:cadcdb5.dao.nrc.ca:4200/cyberska

    DEVSYBASE is the CADC test database.  This will have to be changed to the
    production database when we have permission set up to to direct insertion
    to that resource. Other fields may also change, such as the jdbc driver,
    if production database is not SYBASE (which I do not believe it is).
    For testing purposes, however, I think it is sufficient to use the DEVSYBASE
    just to get Calgary set up and working.


2) wcsLibJNI library Java interface for native datatypes.

   Download opencadc if you do not have it.  Under opencadc/projects find the wcsLibJNI dir.
   Run the make file, which should produce the share object lib,
   libwcsLibJNI.so and write it to opencadc/projects/wcsLibJNI/lib/.
   Once complete. go to the fits2caom directory in DQS/ and
   edit the fits2caom shell script (i.e $DQS/fits2caom/fits2caom) to insert your path
   to libwcsLibJNI.so . You will see my path in there, so comment that out or remove it.


3) $DQS
   Set up the $DQS environment variable.


4) $PYTHONPATH
   Add mddb to the $PYTHONPATH


5) $PATH
    Add dbinsert to $PATH

6)mddb.cfg
  The mddb package uses a configuration file called mddb.cfg to establish the
  operational environment. This file is included under $DQS. Mostly this specifies
  all the directories where files are found, read and written. The naming of the DQS
  dirs should make it obvious what each is for.  Since configs cannot interpret shell
  environment variables, the User MUST edit the first line of this file to indicate
  the DQS directory.

  Eg.,  The Signals Lab $DQS/mddb.cfg file has this at top

  [DEFAULT]
  # UBCO Signals Lab default DQS
  dqs    = /srv/cyberska/DQS      <-- you change this

  Simply edit this to indicate what DQS should be for your particular set up.
  One other config field that you will want to change will be the 'test' switch,
  which is under the 'switches' section.

  [switches]
  verbose = on
  print   = on
  debug   = on
  simple  = on
  no-retrieve = on
  test = on          <-- wait until ready, then flip this to 'off'

  test = on indicates that fits2caom will not execute a database insert, but
  will perform all calculations prior to insertion. This is a good switch to leave
  on until operations appear nominal. When ready for database insertion, simply set
  test = off. Other switches should be left as is. verbose, for example, writes all
  fits2caom gob to a log file, specified witihin the mddb package.

  At this point, things should be ready to go.  mddb provides a test script that
  should indicate nominal operation. Once you have everything configured, run the
  testrun.py script. This is in the scripts/ directory. You can run it from inside
  mddb or outside

  eg.,

  % scripts/testrun.py
  or
  % <path/to/mddb/>scripts/testrun.py

  You should several output lines that looks like this:

  /srv/cyberska/kenneth/sw/mddb> scripts/testrun.py
  ['./dbinsert', '--mtype=image/fits-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=31261', '--hdr=M100contimage.fits.hdr', '--public']
  0x828801395ac068ea


  ['./dbinsert', '--mtype=image/fits-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=31246', '--hdr=Antennae_SouthCO2_1LineCleanimage.fits.hdr', '--public']
  0x5fc901395ac071f0


  ['./dbinsert', '--mtype=image/fits-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=31231', '--hdr=Antennae_NorthCO2_1LineCleanimage.fits.hdr', '--public']
  0x312701395ac07ac0


  ['./dbinsert', '--mtype=image/fits-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=26169', '--hdr=M33-at-160mm.fits.hdr', '--public']
   1

   ['./dbinsert', '--mtype=image/fits-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=24724', '--hdr=atlascdfsfits.fits.hdr', '--public']
   0x52a01395ac08bba


   ['./dbinsert', '--mtype=image/fits-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=17975', '--hdr=chan93_flux.fits.hdr', '--public']
   0x55cc01395ac09469


   ['./dbinsert', '--mtype=image/ms-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=36219', '--hdr=cena.image.tar.hdr', '--public']
   0x626301395ac09d31


   ['./dbinsert', '--mtype=image/ms-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=36223', '--hdr=IRAS16293.ms.tar.hdr', '--public']
   0x540801395ac0a5fd


   ['./dbinsert', '--mtype=image/ms-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=36227', '--hdr=M100cont.ms.tar.hdr', '--public']
   0xa25901395ac0aebd


   ['./dbinsert', '--mtype=image/ms-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=36273', '--hdr=Orion.methanol.ms.tar.hdr', '--public']
   0x1b5401395ac0b774


   ['./dbinsert', '--mtype=image/ms-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=36406', '--hdr=cena_continuum.image.hdr', '--public']
   0xde6601395ac0c044


   ['./dbinsert', '--mtype=image/fits-image', '--uri=http://dms.cyberska.org:8080/dlmanager/getfile?fileid=36899', '--hdr=ALMA_SGR_A.fits.hdr', '--public']
   0x43d401395ac0c90d

NOTE: It is only the testrun.py script that is spitting out the command line arguments.
This is not output of mddb package.

The return codes are seen, either as the fits2caom hex value or 1, which indicates mddb
failure, and no database insertion was attempted. You can view the validation message,
which will appear under $DQS/VALIDATE as the file, M33-at-160mm.fits.valid, which is
just the header file name with the .valid extension replacing .hdr.


D:: Operations

Once the testrun.py script is running and fits2caom is returning hex values
(as opposed to blowing up), you should be good to go for nominal operations.
Hopefully, our validation process will prevent fits2caom blowing up altogether.
As you will see from the testrun.py script, the command line arguments are all
keyword args.

Typing dbinsert (or dbinsert.py) will deliver the Usage:

mddb> dbinsert

      Usage: dbinsert [--help] [--public] [--verbose][--config=<DQS-config-file>] --mtype=<mime-type> --hdr=<hdrfile> --uri=<uri> 

      Three (3) keyword arguments are required. 

      --mtype=		<mime-type> of dataset to be inserted
      	         	One (1) of CyberSKA metadata mime-types:

		     	image/fits-image,
		     	image/ms-image,
		     	image/fits-uvw,
		     	image/ms-uvw

       --hdr=     	<header file for the dataset>
       --uri=     	<URI to CyberSKA platform page, or download link>,
	          	which should look like,

		  	http://www.cyberska.org/pg/file/read/<nnnnn>, or
		  	http://dms.cyberska.org:8080/dlmanager/getfile?fileid=<nnnnnn>

	[--public]  	The dataset metadata are publicly available.
	[--verbose] 	Turn on stdout messages.

	A user may pass a DQS configuration file, i.e., an mddb.cfg:

	[--config=]  	<config-file> path to a DQS config file.
	          	If "--config" is not provided at the command line, a search for "mddb.cfg" will occur on ~ and $DQS.
		  	An Exception is raised if an mddb.cfg file cannot be found.

	[--help]	This message.


1) --config=  At this point, you can see that the command line actually allows you
              to pass an alternative 'mddb.cfg' file, which maybe anywhere. However,
	      I strongly discourage such behaviour. When --config is not specified,
	      the mddb package defaults to looking in the user's home directory (~)
	      and then $DQS, which is where I always leave it and let it run from there.
	      In short, you should never have to pass a --config argument (even though you can).
	      This is provided in the event of multiple DQS directories which may not be the
	      same as specified in the current mddb.cfg file.

2) --hdr=  This is the header file that will come out of the metaData package.

ChangeLog
2012-12-10  Ken Anderson  <kenwood@>
	* mddb v1.3.2 introduces the small updates to modules imports

	* import of mddb packages preprends 'mddb' or 'from mddb'

	* FitsDbInsert.__setSpectralCoord now checks and remediates
	spectral coordinate values of 'FREQUENC" to 'FREQ'

	* README

2012-12-03  Ken Anderson  <kenwood@>

	* v1.3.1 introduces fixes for TELSCOPE and INSTRUME
	keywords, Direction Coordinate checking.

	* DbFitsInsert.__checkSetKeys() method implemented to insert
	'Unknown' for various informational, though db critical, keywords.
	See mddb.db.checkSetKeys for details. In mddb v!3., these
	keys are 'OBJECT', 'INSTRUME', 'TELESCOP', 'OBSERVER'. Default
	values for these keys are provided by checkSetKeys.setFitsKeys.

	* DbFitsInsert.__assertKey() added in support of __checkSetKeys().

	* DbFitsInsert.__setInstrumentLineBasedOnTelescopeValue() added
	in support of DbFitsInsert.__checkSetKeys().

	* db.checkSetKeys provides a database of keyword checks and
	default insertion values.

	* DbFitsInsert.__assertDirectionCoord() looks for a valid direction
	coordinate. Writes ERRORs to validation msg.

	* DbFitsInsert.__affirmDirection() check direction and projection
	codes. Write ERRORs to validation msg.

	* db.checkSetKeys includes direction and projection codes for coordianate
	assertions.

	* DbFitsInsert.__setSpectralCoord() fixes the spectral
	coordinate when the value is incorrectly written as
	'Frequency'. Such a spectral coordinate must be equal to 'FREQ',
	and is so set. A "WARN" entry is logged to the dataset's .valid file.


2012-10-29  Ken Anderson  <kenwood@>

	* ChangeLog: v1.3 introduces a renamed ChangeLog
	(was Changelog)

	* mains package now holds the main class modules:
	baseDbInsert
	dbFitsInsert
	dbCImageInsert
	dbUVFitsInsert
	dbMSInsert

	* dbinsert cli pared.

	* dbinsert now handles NotImplementedError.
	Writes .nogo file to $DQS/VALIDATE.
	Currently, this exception is raised on Mime-types,
	image/fits-uvw, image/ms-uvw

	* dbinsert.run() function refactored to utils.rUtils,
	i.e. utils.rUtils.run()

	* DbFitsInsert.configureOverride() method now traps/reports
	FitsHeaderError exceptions in .valid files as an ERROR.

	* DBFitsInsert implements new method, __assertSimple(),
	testing for MEF files. MEF files are not handled.
	ERROR is written to validationMsg. ERROR prevents
	attempt at executeInsert() call.

	* confirmKeys.requireFitsKeys map updated and expanded; now
	includes WCS keywords.

	* DbUVFitsInsert.configureOverride() defers to baseDbInsert
	to raise NotImplementedError

	* DbMSInsert.configureOverride() defers to baseDbInsert
	to raise NotImplementedError

2012-10-24  Ken Anderson  <kenwood@>

	* __dates2mjd() now premeptively looks for MJD-OBS keyword.
	No action and return if found.

2012-10-14  aVakt  <vakt@aVakt.local>

	* ChangeLog: v1.2 introduces a number of minor bug fixes from 1.0, 1.1

	* Error status now returns -1 on VALIDATION errors.

	* velrefMap updated to include mappings of single digits encountered in
	in FITS VELREF keywords. Single digits are not nominal values, but are
	now handled.

	* dbinsert call on dbMsInsert() fixed (is dbMSInsert())

	* constructor interface fixed for NotImplemented classes,
	dbMSInsert, dbUVFitsInsert

	* Fixed str concat exception in executeInsert() (float ndims
	adding to str)


2012-10-08  aVakt  <vakt@aVakt.local>

	* ChangeLog: v1.1 introduces minor updates to version 1.0

	* '--nodb' A cli switch that exits before db insertion.
	Exits prior to the call on executeInsert().
	This is alterated in dbinsert.py and utils.rUtils

	* 'ENGINEERING' keyword added to the .override file;
	indicates mddb package name, version.




Copyright 1997, 1998 Free Software Foundation, Inc.
Copying and distribution of this file, with or without modification, are
permitted provided the copyright notice and this notice are preserved.


utils/rUtils.py
#!/usr/bin/env python
#
#                                                           CyberSKA DQS Project
#
#                                                           mddb.utils.rUtils.py
#                                                      Kenneth Anderson, 2012-03
#                                                            ken.anderson@ubc.ca
# ------------------------------------------------------------------------------

# $Id$
# ------------------------------------------------------------------------------
__version__      = '$Revision$'[11:-3]
__version_date__ = '$Date$'[7:-3]
__author__       = "k.r. anderson, <ken.anderson@ubc.ca>"
# ------------------------------------------------------------------------------

"""Some utility funcs for command line interface."""
# ------------------------------------------------------------------------------

import sys
import getopt
import urlparse
from   datetime import date

from os import getenv
from os.path import basename, join, isfile, expanduser, split, splitext

from mddb import mddbVersion

from mddb.mains import dbFitsInsert
from mddb.mains import dbCImageInsert
from mddb.mains import dbUVFitsInsert # -- future implementation, raise NotImplementError
from mddb.mains import dbMSInsert     # -- future implementation, raise NotImplementError

# ------------------------------------------------------------------------------

def usage(mod):
    useBurp = '\n\t'+mod+' running '+mddbVersion.pkg_name+' v'\
        +mddbVersion.version +\
        '\n\n\tUsage: '+ mod + ' '\
        '[--help] ' \
        '[--public] ' \
        '[--verbose]' \
        '[--nodb]' \
        '[--config=<DQS-config-file>] ' \
        '--mtype=<mime-type> ' \
        '--hdr=<hdrfile> --uri=<uri> \n\n\t' \
        'Three (3) keyword arguments are required. \n\n\t'\
        '--mtype= \t<mime-type> of dataset to be inserted\n\t\t'\
                 '\tOne (1) of CyberSKA metadata mime-types:\n\n\t\t'\
                 '\t\timage/fits-image, \n\t\t'\
                 '\t\timage/ms-image, \n\t\t'\
                 '\t\timage/fits-uvw,\n\t\t'\
                 '\t\timage/ms-uvw\n\n\t'\
        '--hdr=   \t<header file for the dataset>\n\t'\
        '--uri=   \t<URI to CyberSKA platform page, or download link>,\n\t\t'\
                 '\twhich should look like,\n\n\t\t'\
                 '\thttp://www.cyberska.org/pg/file/read/<nnnnn>, or\n\t\t'\
                 '\thttp://dms.cyberska.org:8080/dlmanager/getfile?fileid=<nnnnnn>\n\n\t'\
        '--public\tThe dataset metadata are publicly available.\n\t'\
        '--verbose\tTurn on stdout messages.\n\t' \
        '--nodb\t\tNo DB insert. executeInsert() is not called.\n\n\t'\
        'A user may pass a DQS configuration file, i.e., an mddb.cfg:\n\n\t'\
        '--config= \t<config-file> path to a DQS config file.\n\n\t'\
        '\t\tIf "--config" is not provided at the command line, a search \n\t'\
        '\t\tfor "mddb.cfg" will occur on ~ and $DQS.  An Exception is\n\t'\
        '\t\traised if an mddb.cfg file cannot be found.\n\n\t'\
        '--help \t\tThis message.\n\n'
    return useBurp


def handleCLargs(args):
    mod = basename(sys.argv[0])
    long_options = ['help','public','verbose', 'nodb','mtype=','hdr=','uri=','config=']
    required     = ['--mtype', '--hdr', '--uri']
    nrequired    = len(required)
    Nreqd        = 0
    valid_mtypes = ['image/fits-image',
                    'image/ms-image',
                    'image/fits-uvw',
                    'image/ms-uvw'
                    ]
    try:
        opts, arg = getopt.getopt(sys.argv[1:],'',long_options)
    except getopt.GetoptError,err:
        print "Option parsing recieved an error:"
        print err
        sys.exit(usage(mod))
        
    if arg:
        print "Unexpected arg(s):",arg
	sys.exit(usage(mod))

    # Three (3) arguments are required, see 'required' list. 
    # --mtype   mime-type of the dataset to be inserted
    # --hdr     header file for the dataset
    # --uri     URI to the CyberSKA file platform page,
    #           which hould look like,
    #           www.cyberska.org/pg/file/read/<nnnnn>/<page-name>
    # --public  indicates that the dataset is publicly accessible.
    # --verbose Turns on stdout messages.
    # --config  may be passed. This is a path to an mddb.cfg file.
    # --nodb    No DB interaction; executeInsert() is not called.

    cl_args = []
    if opts:
	for o, a in opts:
	    if o in ("--help",):
		sys.exit(usage(mod))
            if o in ("--public"):
                cl_args.append(o)
            if o in ("--verbose"):
                cl_args.append(o)
            if o in ("--nodb"):
                cl_args.append(o)
	    if a and o in required:
                Nreqd +=1
		cl_args.append(o+"="+a)
            elif a:
                cl_args.append(o+"="+a)
    else: sys.exit(usage(mod))

    # Ensure the required three (3) args have been passed.
    if Nreqd != nrequired:
        print "\n\tMissing required arguments (see usage)."
        sys.exit(usage(mod))

    # Ensure valid CyberSKA mime-type.
    err = "\nError:: Unrecognized mime-type:"
    for kwarg in cl_args:
        if '--mtype' in kwarg:
            key,val = kwarg.split('=')
            if val not in valid_mtypes:
                print err,val
                sys.exit(usage(mod))
        else: continue
    return cl_args


def parseArgs(clargs):
    pargs = {}
    pargs['public']  = False    # permission default is private.
    pargs['verbose'] = False    # default no stdout messaging
    pargs['nodb']    = False    # default runs executeInsert()

    for kwarg in clargs:
        if '--public' in kwarg:
            pargs['public'] = True
            continue
        if '--verbose' in kwarg:
            pargs['verbose'] = True
            continue
        if '--nodb' in kwarg:
            pargs['nodb'] = True
            continue
        if '--uri' in kwarg:
            urlset = urlparse.urlparse(kwarg)
            key,val = (urlset.path).split('=')
            val= val+"?"+urlset.query
            pargs[key.strip('--')] = val
            continue
        else:
            key,val = kwarg.split('=')
            pargs[key.strip('--')] = val

    if 'config' in pargs:
        cfgFile = expanduser(pargs['config'])
        if isfile(cfgFile): pargs['config'] = cfgFile
        else: raise RuntimeError(pargs['config']+" Not found.")
    else:
        cfgFile = findCfg()
        pargs['config'] = cfgFile
    return pargs


def findCfg():
    """Hunt down an 'mddb.cfg' file, first in ~, then in $DQS. Raise expcetion
    if none found.
    
    parameters: <void>
    return:     <string>, mddb.cfg path
    """
    noEnvVar     = 'mddb.cfg not found in ~; DQS environment variable not defined.'
    filenotfound = 'mddb configuration file NOT found. Check ~ and $DQS'

    cfgFile  = None
    testhome = join(getenv('HOME'),'mddb.cfg')

    if isfile(testhome):
        cfgFile = testhome
    else:
        try:
            dqsFile = join(getenv('DQS'),'mddb.cfg')
            if not isfile(dqsFile):
                raise RuntimeError(filenotfound)
        except AttributeError:
            raise RuntimeError(noEnvVar)
        if isfile(dqsFile):
            cfgFile = dqsFile
        else:
            raise RuntimeError(filenotfound)
    return cfgFile


def getCollId(uri):
    """Get the CyberSKA file_guid id or a fileid value from a passed URI string.
    A cyberska.org URI will be of the form,

    http://www.cyberska.org/pg/file/read/<nnnnn>/<page-name>,       or
    http://dms.cyberska.org:8080/dlmanager/getfile?fileid=<nnnnn>
    
    where <nnnnn> is a file_guid ID or fileid

    parameters: <string>, cyberska.org URI string
    return:     <string>, ID value
    """
    purl = urlparse.urlparse(uri)
    if purl.query:
        try:
            collid = urlparse.parse_qs(purl.query)['fileid'][0]
        except AttributeError:
            collid = purl.query.split('=')[1]
    else:
        collid = split(purl.path)[-1]
    return collid


# v1.1 added '--nodb' switch, halts processing, does not call 
# executeInsert()

def run(clArgs):
    useFits    = useUVFits = useCimage = useUVMSet = False
    pArgs      = parseArgs(clArgs)
    uri        = pArgs['uri']
    hdrFile    = pArgs['hdr']
    fileType   = pArgs['mtype']
    configFile = pArgs['config']
    public     = pArgs['public']
    verbose    = pArgs['verbose']
    nodbInsert = pArgs['nodb']
    collid     = getCollId(uri)
    abortString="Database insert aborted. See validation file."

    # Determine class use by fileType
    if fileType == 'image/fits-image':
        useFits   = True
    elif fileType == 'image/ms-image':
        useCimage = True
    elif fileType == 'image/fits-uvw':
        useUVFits = True
    elif fileType == 'image/ms-uvw':
        useUVMSet = True

    if verbose:
        print "\n\n\tThis is mddb, v"+mddbVersion.version
        print "\t"+("-")*19+"\n"
        print "\n\nConfiguration parameters:"
        print "__________________"
        print "DB config file:\t",configFile
        print "DB coll ID:\t",collid
        print "Header file:\t",hdrFile
        print "MIME-type:\t",fileType
        print "Resource URI:\t",uri
        print "Public Release:\t",public
        print "__________________\n"


    if useFits:
        dbtask  = dbFitsInsert.DbFitsInsert(configFile, verbose)
        override= dbtask.buildOverride(hdrFile, uri, public)
        msgTypes= dbtask.confirmOverride(override, fileType)
        if nodbInsert:
            sys.exit("No DB request. Done")
        elif not "ERROR:" in msgTypes: 
            dbtask.executeInsert(collid, override, uri)
        else:
            if verbose: print abortString
            sys.exit("-1")
    elif useCimage:
        dbtask  = dbCImageInsert.DbCImageInsert(configFile, verbose)
        override= dbtask.buildOverride(hdrFile, uri, public)
        msgTypes= dbtask.confirmOverride(override, fileType)
        if nodbInsert:
            sys.exit("No DB request. Done")
        elif not "ERROR:" in msgTypes:
            dbtask.executeInsert(collid, override, uri)
        else:
            if verbose: print abortString
            sys.exit("-1")
    elif useUVFits:
        dbtask  = dbUVFitsInsert.DbUVFitsInsert(configFile, verbose)
        try: override= dbtask.buildOverride(hdrFile,uri,public)
        except NotImplementedError, err:
            writeNoGo(hdrFile,fileType,dbtask.dbEnv.VALIDATE)
            sys.exit("-2")
        msgTypes= dbtask.confirmOverride(override,fileType)
        if nodbInsert:
            sys.exit("No DB request. Done")
        elif not "ERROR:" in msgTypes:
            dbtask.executeInsert(collid,override,uri)
        else:
            if verbose: print abortString
            sys.exit("-1")
    elif useUVMSet:
        dbtask  = dbMSInsert.DbMSInsert(configFile, verbose)
        try: override= dbtask.buildOverride(hdrFile,uri,public)
        except NotImplementedError, err:
            writeNoGo(hdrFile,fileType,dbtask.dbEnv.VALIDATE)
            sys.exit("-2")
        msgTypes= dbtask.confirmOverride(override,fileType)
        if nodbInsert:
            sys.exit("No DB request. Done")
        elif not "ERROR:" in msgTypes:
            dbtask.executeInsert(collid,override,uri)
        else:
            if verbose: print abortString
            sys.exit("-1")
    return


def writeNoGo(hdr, mtype, where):
    """Write a .nogo file when NotImplementedError is raised.

    parameters: <string>, <string, <string>
    return:     <void>
    """

    path,hfile=split(hdr)
    head = splitext(split(hdr)[1])[0]
    nogoFile = join(where,head)+".nogo"

    nogo = open(nogoFile,"w")
    nogo.write("# MDDB Nogo file written for header file: "+hdr+"\n")
    nogo.write("# Written "+date.today().isoformat()+"\n")
    nogo.write("# Package Name: "+mddbVersion.pkg_name+"\n")
    nogo.write("# Package Version: "+mddbVersion.version+"\n\n")
    nogo.write("File, "+hdr+", is a nogo for CAOM v1 database insertion.\n\n")
    nogo.write("Received a NotImplementedError\n")
    nogo.write("Unsupported MIME-TYPE: "+mtype)
    return
